fixing socket

G2G2G2G opened the pull request:
Firefox will trigger `write error: websocket: close sent` (chromium seems to be unable to trigger this event, even with thousands of refreshes sent) by pressing refresh several times on a page with a load time of (network latency included) >250ms 
This causes the `unregister <- connection` to fire, however this breaks the socket entirely for ALL current connections and all future connections will connect to the websocket but the websocket is down and will not transfer / receive messages. (Messages can be sent to the server ACCORDING TO THE BROWSER but without error but they're thrown out by the server without notice of any kind).

No idea why it happens like that, but it does. Posted fix works fine.

kiyonlin commented on websocket-chat/main.go:
```diff
@@ -30,8 +30,9 @@ func runHub() {
-					connection.WriteMessage(websocket.CloseMessage, []byte{})
+					//unregister <- connection
+					//connection.WriteMessage(websocket.CloseMessage, []byte{})
+					delete(clients, connection)
```
You can't delete it directly which will cause race condition(read and write clients map at the same time)

kiyonlin commented:
> Firefox will trigger write error: websocket: close sent by pressing refresh 

This is an expected result. But Firefox and chromium may have different strategy to deal with the websocket connection. I guess Firefox closes the connection and chromium keeps it. Although i'm not very sure about why they have different behaviors, but I'm sure the code in this recipe doesn't have issue.

G2G2G2G commented:
Well it crashes the entire program with a panic, so that's obviously a problem with the code since people are going to be dealing with that on their server... There's a reason why fasthttp & gorilla's examples are drastically different.

kiyonlin commented:
> Well it crashes the entire program with a panic, so that's obviously a problem with the code since people are going to be dealing with that on their server... There's a reason why fasthttp & gorilla's examples are drastically different. 
> 
> 
> 
> If my fix isn't good (even though it works in production past few months while the recipe doesn't even work for 24 hours) delete the recipe entirely and copy fasthttps example instead. 

I think you are right, will take a look and run it tomorrow.

kiyonlin commented on websocket-chat/main.go:
```diff
@@ -30,8 +30,9 @@ func runHub() {
-					unregister <- connection
-					connection.WriteMessage(websocket.CloseMessage, []byte{})
+					//unregister <- connection
+					//connection.WriteMessage(websocket.CloseMessage, []byte{})
```
I think we still need to send close message and close the connection, and then we delete it from map.

G2G2G2G commented:
Is there a way to connection.WriteMessage( without crashing the program? Cuz firefox closes the connection and Go crashes when it fails to write that message (because the connection is closed but the program still tries to write to that connection) The browser already knows the socket is dead, before that message is sent regardless... I have no idea what to do there, all I know is what I have tested and I ended up having to go to this: https://github.com/fasthttp/websocket/tree/master/_examples/chat and convert it to fiber but I still had to delete this line 94 https://github.com/fasthttp/websocket/blob/master/_examples/chat/client.go#L94 
Every browser mobile and desktop seems to not have any issue and knows when the socket is closed immediately. My employees have apple, android, linux desktops, windows, OS X, etc it works on all of them but when that line 94 is there it crashes the entire program with ` panic: send on closed channel`
I do not understand why it does this or anything, I just know it does do it.
On my socket in the readPump I take a switch statement and run a function with a query to send back database input. This makes my write messages have like 10ms delay vs the <1ms delay of normal. The only reason these examples seem to work is because the computers they're ran on respond "fast enough" that isn't good.. if the program is ran on a slow computer or more code is added and it is too slow then your program starts to crash. It shouldn't be a requirement to run it on a fast enough connection/computer without crashing. Also happens when you throttle network latency to like 300ms

kiyonlin commented:
Thanks for the details. I need to dive into the source code to see what happened when got error while writing message.

kiyonlin commented:
> when that line 94 is there it crashes the entire program with `panic: send on closed channel`

Could you show us the total message you got when the program crashed?

G2G2G2G commented:
> > when that line 94 is there it crashes the entire program with `panic: send on closed channel`
> 
> Could you show us the total message you got when the program crashed?

I changed my entire program to follow the fasthttp example (but sends the fiber session stuff to it) one but I can still trigger the same error message.
I just wrote that message in my notes to myself.
```				
//commented out because causes error:  panic: send on closed channel
//c.WriteMessage(websocket.CloseMessage, []byte{})
```

I actually got it to trigger via chromium this time and it was when I refreshed a page with several queries on it that send via the websocket and then immediately closed chromium (alt+f4 basically)

```Bash
Apr 15 11:31:52 manager-websrv manager[28994]: **panic: send on closed channel**
Apr 15 11:31:52 manager-websrv manager[28994]: goroutine 283 [running]:
Apr 15 11:31:52 manager-websrv manager[28994]: main.socReader(0x2b52540)
Apr 15 11:31:52 manager-websrv manager[28994]:         /projects/go/manager/websocket.go:166 +0xd34
Apr 15 11:31:52 manager-websrv manager[28994]: created by main.manageSocket
Apr 15 11:31:52 manager-websrv manager[28994]:         /projects/go/manager/websocket.go:103 +0x154
```
The 2 lines it says are 103 and 166
103 calls my "readPump" which I call `go socReader(client)` 
```Go
func manageSocket(c *websocket.Conn, h *Hub) {
	client := &Client{hub: h, conn: c, send: make(chan []byte, 256)}
	client.hub.register <- client
	pingticker := time.NewTicker(45 * time.Second)
	defer func() {
		pingticker.Stop()
		client.hub.unregister <- client
		err := c.Close()
		if err != nil {
			log.Println("c.close:", err)
		}
	}()
	go socReader(client)
```
(like I said I basically coped and changed it to work the fasthttp one) 
and line 166 is a normal message:
`c.send <- getProjects(id, est)`

**I am right when I think that even if you do really stupid stuff on the client you shouldn't be able to crash the server right?**
The fasthttp one is way harder to crash, the old code I had that I changed back in february was much easier.. is me treating the browser so poorly not a good excuse? I figure no matter what someone messing with your site shouldn't be able to crash the entire thing. Also maybe I combined fiber + fasthttp example incorrectly but it has been working for months with all the employees using it just fine. (I don't really know what I am doing though lol)

G2G2G2G commented:
Actually one of my employees said they can cause it to crash just spamming F5 on any page. Error is the same error `panic: send on closed channel`
IDK what their computer has that is different, maybe they have higher latency than I do to the server. 
https://github.com/fasthttp/websocket/blob/master/_examples/chat/client.go#L88
How do I change this to not panic?
This is my version of it, almost identical:
```Go
for {
		select {
		case message, ok := <-client.send:
			c.SetWriteDeadline(time.Now().Add(10 * time.Second))
			if !ok {
				// The hub closed the channel.

				//comment out this might cause error:  panic: send on closed channel
				//c.WriteMessage(websocket.CloseMessage, []byte{})

				return
			}

			w, err := c.NextWriter(websocket.TextMessage)
			if err != nil {
				return
			}
			w.Write(message)

			// Add queued chat messages to the current websocket message.
			n := len(client.send)
			for i := 0; i < n; i++ {
				//change the thing to be parsable because \n can be part of JSON but -\n- is never? going to happen...
				//could still happen though, but lots of stuff has \n from textareas so it is bad to use only "\n"
				w.Write([]byte("-\n-"))
				w.Write(<-client.send)
			}

			if err := w.Close(); err != nil {
				return
			}
		case <-pingticker.C:
			c.SetWriteDeadline(time.Now().Add(10 * time.Second))
			if err := c.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		}
	}
```
I don't really understand what all the queue stuff is, it does w.Write(<-client.send) instead of c.WriteMessage.. but both fiber and fasthttp panic due to the same reason, I guess it's a fasthttp issue? It might be a gorilla socket issue too

kiyonlin commented:
> Error is the same error panic: `send on closed channel`

I kind of know why does this happen. 

I guess you did this `c.send <- getProjects(id, est)` in another goroutine function, but that function didn't know the `c.send` channel was closed(go disallow sending data to a closed channel, will cause panic). So I suggest you to add `defer func(){ recover()}()` at the beginning of that goroutine function.

G2G2G2G commented:
https://github.com/fasthttp/websocket/blob/master/_examples/chat/client.go#L135
This is in multiple go routines, I did the same and both crash on that 
Where do you suggest putting the defer? is that how this works in my main loop?
`	manage.Use(recover.New())`
I have that already that will go to the `defer func(){ recover()}()` you say?

kiyonlin commented:
Put it on all functions which contain `c.send<-"message"` like 
```go
func func1() {
	defer func() { recover() }()
	//...
	c.send <- "message"
	//...
}
```

G2G2G2G commented:
Thanks I think that worked, I only have 1 that does it and the hub, should I add it to the hub as well?

kiyonlin commented:
> Thanks I think that worked, I only have 1 that does it and the hub, should I add it to the hub as well?
> https://github.com/fasthttp/websocket/blob/master/_examples/chat/hub.go#L32
> this one

If the function doesn't know if the channel is closed, it should be protected by `recover()`.

G2G2G2G commented:
Ok I wasn't sure since hub is a little different and ran from here https://github.com/fasthttp/websocket/blob/master/_examples/chat/main.go#L31 but yea it does send to the channel as you said so I'll add it to hub, we also tried to crash it and can't now so it is looking good, had like 5 people in different locations all trying at the same time. Before we could crash it in like 5 seconds if we tried hard.

Thank you for the help!

I am not sure what to do should we just add a recover in the example then?
https://github.com/gofiber/recipes/blob/master/websocket-chat/main.go#L60
The hub also is separate and has no defer inside of it.

kiyonlin commented:
The example doesn't have the `send on closed channel` issue, but it will get deadlock while calling `unregister <- connection` cuz unregister is a blocked channel. 
I think we should use `delete(clients, connection)` instead of it.

kiyonlin commented:
The final code will be 
```go
connection.WriteMessage(websocket.CloseMessage, []byte{})
connection.Close()
delete(clients, connection)
```

Would you like to finish it?

G2G2G2G commented:
I'm not sure how to do it I don't really use github, I just report issues if I see them lol
You can do it. Thanks for all your help!

kiyonlin commented:
Okay thanks for reporting! And I'll fix it.
